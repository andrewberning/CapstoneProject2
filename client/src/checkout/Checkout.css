.Checkout {
  max-width: 400px;
  margin: 0 auto;
  text-align: center;
}

.checkout-container {
  background-color: #fff;
  border-radius: 8px;
}

.checkout-details p {
  font-size: 1.1rem;
}

#paypal-button-container {
  margin-top: 20px;
  display: flex;
  justify-content: center;
}

.error-message {
  text-align: center;
  margin-bottom: 20px;
}
