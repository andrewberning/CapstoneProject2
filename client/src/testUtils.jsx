import React from "react";
import UserContext from "./auth/UserContext";
import { vi } from 'vitest';

const demoUser = {
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "testuser123@test.net",
};

// Mock functions
const mockSetCurrentUser = vi.fn();
const mockLogin = vi.fn();
const mockLogout = vi.fn();
const mockSignup = vi.fn();
const mockAddToCart = vi.fn();
const mockHasAddedToCart = vi.fn();

const UserProvider = ({ 
  children,
  currentUser = demoUser,
  setCurrentUser = mockSetCurrentUser,
  login = mockLogin,
  logout = mockLogout,
  signup = mockSignup,
  addToCart = mockAddToCart,
  hasAddedToCart = mockHasAddedToCart,
  cartItems = [],
  setCartItems = () => {},
  categories = [],
  infoLoaded = true,
  totalAmount = 0,
  totalItems = 0,
}) => (
  <UserContext.Provider value={{ 
    currentUser,
    setCurrentUser,
    login,
    logout,
    signup,
    addToCart,
    hasAddedToCart,
    cartItems,
    setCartItems,
    categories,
    infoLoaded,
    totalAmount,
    totalItems,
  }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider, mockSetCurrentUser, mockLogin, mockLogout, mockSignup, mockAddToCart, mockHasAddedToCart };
