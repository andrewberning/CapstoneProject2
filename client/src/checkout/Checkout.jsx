import { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import UserContext from "../auth/UserContext";
import { PayPalButtons, PayPalScriptProvider } from "@paypal/react-paypal-js";
import ShoplyApi from "../api/api";
import "./Checkout.css";

export default function Checkout() {
  const { cartItems, setCartItems, currentUser, totalAmount, totalItems } = useContext(UserContext);
  const navigate = useNavigate();
  const [message, setMessage] = useState("");

  const initialOptions = {
   "client-id": "AfBVLM9djpLsUHVucMM1dygLEeRCzucMls_rlkLkbNCF_tQpLblElMcEfwCrXJGfeZyNODX4uSunvIw4",
    currency: "USD",
    intent: "capture",
  };

  const createOrder = async () => {
    try {
      const response = await ShoplyApi.createPayPalOrder({
        cart: cartItems,
        totalAmount: totalAmount,
        totalItems: totalItems,
      })

      const orderData = response;

      if (orderData.jsonResponse.id) {
        return orderData.jsonResponse.id;
      } else {
        const errorDetail = orderData?.details?.[0];
        const errorMessage = errorDetail
          ? `${errorDetail.issue} ${errorDetail.description} (${orderData.debug_id})`
          : JSON.stringify(orderData);

        throw new Error(errorMessage);
      }
    } catch (error) {
      console.error(error);
      setMessage(`Could not initiate PayPal Checkout...${error}`);
    }
  };

  const onApprove = async (data, actions) => {
    try {
      const response = await ShoplyApi.capturePayPalOrder(data.orderID);

      const orderData = response;
      const errorDetail = orderData?.details?.[0];

      if (errorDetail?.issue === "INSTRUMENT_DECLINED") {
        return actions.restart();
      } else if (errorDetail) {
        throw new Error(
          `${errorDetail.description} (${orderData.debug_id})`
        );
      } else {
        const transaction = orderData.purchase_units[0].payments.captures[0];
        setMessage(
          `Transaction ${transaction.status}: ${transaction.id}. See console for all available details`
        );
        console.log(
          "Capture result",
          orderData,
          JSON.stringify(orderData, null, 2)
        );

        // Create an order in database

        const shippingInfo = orderData.purchase_units[0].shipping;

        await ShoplyApi.createOrder({
          userId: currentUser.id,
          totalItems,
          totalAmount,
          cartItems,
          name: shippingInfo.name.full_name,
          address: shippingInfo.address.address_line_1,
          city: shippingInfo.address.admin_area_2,
          state: shippingInfo.address.admin_area_1,
          zip: shippingInfo.address.postal_code,
          country: shippingInfo.address.country_code,
        });

        // delete items from cart in database
        await ShoplyApi.removeAllItemsFromCart(currentUser.cartId)
        setCartItems([]);
        
        // navigate to order confirmation page
        navigate('/order-confirmation');
      }
    } catch (error) {
      console.error(error);
      setMessage(
        `Sorry, your transaction could not be processed...${error}`
      );
    }
  };

  

  return (
    <div className="Checkout my-4">
      <PayPalScriptProvider options={initialOptions}>
        <h2 className="text-center mb-4">Checkout</h2>
        {message && <p className="error-message alert alert-danger">{message}</p>}
        <div className="checkout-container card p-4 shadow-sm">
          <div className="checkout-details">
            <p className="mb-4">Items ({totalItems}): <strong>${totalAmount.toFixed(2)}</strong></p>
            <div id="paypal-button-container">
              <PayPalButtons 
                style={{
                  shape: "pill",
                  layout: "vertical",
                  color: "gold",
                  label: "paypal",
                }}
                createOrder={createOrder}
                onApprove={onApprove}
              />
            </div>
          </div>
        </div>
      </PayPalScriptProvider>
    </div>
  );

}