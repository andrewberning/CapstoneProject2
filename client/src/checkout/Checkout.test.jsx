import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Checkout from "./Checkout";
import { UserProvider } from "../testUtils";

describe('Checkout component', () => {
  it('renders without crashing', () => {
    render(
      <MemoryRouter>
        <UserProvider >
          <Checkout />
        </UserProvider>
      </MemoryRouter>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserProvider>
          <Checkout />
        </UserProvider>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});