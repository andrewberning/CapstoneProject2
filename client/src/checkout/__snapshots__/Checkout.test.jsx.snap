// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkout component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Checkout my-4"
  >
    <h2
      class="text-center mb-4"
    >
      Checkout
    </h2>
    <div
      class="checkout-container card p-4 shadow-sm"
    >
      <div
        class="checkout-details"
      >
        <p
          class="mb-4"
        >
          Items (0): 
          <strong>
            $0.00
          </strong>
        </p>
        <div
          id="paypal-button-container"
        >
          <div
            class=""
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
